src/psbt.c

```diff
@@ -0,0 +1,2120 @@
+    int ret;
+    size_t i;
+    bool found_sep = false;
+
+    // Init and alloc the maps
+    if (counts.num_keypaths > 0) {
+        if ((ret = keypath_map_init_alloc(counts.num_keypaths, &result->keypaths)) != WALLY_OK) {
+            return ret;
+        }
+    }
+    if (counts.num_partial_sigs > 0) {
+        if ((ret = partial_sigs_map_init_alloc(counts.num_partial_sigs, &result->partial_sigs)) != WALLY_OK) {
+            return ret;
+        }
+    }
+    if (counts.num_unknowns > 0) {
+        if ((ret = unknowns_map_init_alloc(counts.num_unknowns, &result->unknowns)) != WALLY_OK) {
+            return ret;
+        }
+    }
+
+    // Read key value pairs
+    while (p < end) {
+        // Read the key length
+        p += varint_from_bytes(p, &key_len);
+
+        if (key_len == 0) {
+            found_sep = true;
+            break;
+        }
+
+        // Read the key itself
+        key = p;
+        type = key[0];
+        p += key_len;
+
+        // Process based on type
+        switch (type) {
+        case WALLY_PSBT_IN_NON_WITNESS_UTXO: {
+            if (result->non_witness_utxo) {
+                return WALLY_EINVAL;     // We already have a non witness utxo
+            } else if (key_len != 1) {
+                return WALLY_EINVAL;     // Global tx key is one byte type
+            }
+            p += varint_from_bytes(p, &value_len);
+            value = p;
+            wally_tx_from_bytes(value, value_len, 0, &result->non_witness_utxo);
+            p += value_len;
+            break;
+        }
+        case WALLY_PSBT_IN_WITNESS_UTXO: {
+            uint64_t amount = -1, script_len;
+            size_t script_len_len;
+            if (result->witness_utxo) {
+                return WALLY_EINVAL;     // We already have a witness utxo
+            } else if (key_len != 1) {
+                return WALLY_EINVAL;     // Global tx key is one byte type
+            }
+            p += varint_from_bytes(p, &value_len);
+            p += uint64_from_le_bytes(p, &amount);
+            script_len_len = varint_from_bytes(p, &script_len);
+            p += script_len_len;
+            ret = wally_tx_output_init_alloc(amount, p, script_len, &result->witness_utxo);
+            if (ret != WALLY_OK) {
+                return ret;
+            }
+            p += script_len;
+            if (8 + script_len_len + script_len != value_len) {
+                return WALLY_EINVAL;
+            }
+            break;
+        }
+        case WALLY_PSBT_IN_PARTIAL_SIG: {
+            struct partial_sigs_map *partial_sigs = result->partial_sigs;
+            if (key_len != 66 && key_len != 34) {
+                return WALLY_EINVAL;     // Size of key is unexpected
+            }
+            // Check for duplicates
+            for (i = 0; i < partial_sigs->num_items; ++i) {
+                if (memcmp(partial_sigs->items[i].pubkey, &key[1], key_len - 1) == 0) {
+                    return WALLY_EINVAL;     // Duplicate key
```

Comment:
possible leaks of all the memory allocated above it

Reply:
I don't think so. All of the memory allocated is part of the psbt which is passed in by the caller. If these from_bytes functions fail, then the actual `wally_psbt_from_bytes` function will free the entire psbt and thus any memory that was allocated here.

Reply:
ok